lib/web/desktop_playback.go

```diff
@@ -0,0 +1,248 @@
+				// Only a warning as we expect this case if the websocket is
+				// closed by another goroutine (net.ErrClosed) or by the browser (io.EOF)
+				// while websocket.JSON.Receive() is hanging.
+				pp.log.WithError(err).Warn("error reading from websocket")
+			} else {
+				pp.log.WithError(err).Error("error reading from websocket")
+			}
+			return
+		}
+		pp.log.Debugf("recieved playback action: %+v", action)
+		if action.Action == actionPlayPause {
+			pp.togglePlaying()
+		} else {
+			pp.log.Errorf("received unknown action: %v", action.Action)
+			return
+		}
+	}
+}
+
+// StreamSessionEvents streams the session's events as playback events over the websocket.
+func (pp *playbackPlayer) StreamSessionEvents() {
+	defer pp.log.Debug("playbackPlayer.StreamSessionEvents returned")
+	defer pp.close()
+
+	var lastDelay int64
+	eventsC, errC := pp.clt.StreamSessionEvents(pp.ctx, session.ID(pp.sID), 0)
+	for {
+		pp.hangWhilePaused()
+
+		select {
+		case err := <-errC:
+			if errors.Is(err, context.Canceled) {
+				// Expected if the request context is cancelled.
+				pp.log.WithError(err).Warningf("streaming session %v", pp.sID)
+			} else {
+				pp.log.WithError(err).Errorf("streaming session %v", pp.sID)
+			}
+			return
+		case evt := <-eventsC:
+			if evt == nil {
+				pp.log.Debug("reached end of playback")
+				// deferred ps.Close() will set ps.playState = playStateFinished for us
+				return
+			}
+			switch e := evt.(type) {
+			case *apievents.DesktopRecording:
+				if e.DelayMilliseconds > lastDelay {
+					time.Sleep(time.Duration(e.DelayMilliseconds-lastDelay) * time.Millisecond)
+					lastDelay = e.DelayMilliseconds
+				}
+				if _, err := pp.ws.Write(e.Message); err != nil {
+					if errors.Is(err, net.ErrClosed) {
+						// Only a warning as we expect this case to arise when another
+						// goroutine returns before this one or the browser window is closed,
+						// both of which cause the websocket to close.
+						pp.log.WithError(err).Warn("failed to write TDP message over websocket")
+					} else {
+						pp.log.WithError(err).Error("failed to write TDP message over websocket")
+					}
+					return
+				}
+			default:
+				pp.log.Warnf("session %v contains unexpected event type %T", pp.sID, evt)
+			}
+		}
+	}
+}
+
+// Wait waits until the playbackPlayer's context is cancelled.
+func (pp *playbackPlayer) Wait() {
+	defer pp.log.Debug("playbackPlayer.Wait returned")
+	defer pp.close()
+	<-pp.ctx.Done()
+	return
+}
+
+func (h *Handler) desktopPlaybackHandle(
+	w http.ResponseWriter,
+	r *http.Request,
+	p httprouter.Params,
+	ctx *SessionContext,
+) (interface{}, error) {
+	sID := p.ByName("sid")
+	if sID == "" {
+		return nil, trace.BadParameter("missing sid in request URL")
+	}
+
+	websocket.Handler(func(ws *websocket.Conn) {
+		defer h.log.Debug("desktopPlaybackHandle websocket handler returned")
+
+		pp := NewPlaybackPlayer(sID, ws, ctx.clt, r.Context(), h.log)
```

Comment:
It might be a little cleaner to not pass the context to new, and instead pass it as the first argument to `ReceiveActions` and `StreamSessionEvents`.

Context should usually be passed as an argument and rarely "stored" in a struct. Plus, in this case, `New` is not the long running operation that needs to support cancelation, `ReceiveActions` and `StreamSessionEvents` are.

(We tend to store context in structs all the time in this code base, so there are exceptions, but I'm trying to get us away from this pattern.)